### PR TITLE
Fee payments hash in SidechainBlock

### PR DESCRIPTION
**Pull request by i-Alex**

### CODE REVIEW

File: qa/websocket_server_fee_payments.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        ws.checkMessageStaticFields(new_tip_event,
+                                    ws.EVENT_MSG_TYPE,
+                                    -1,
+                                    ws.UPDATE_TIP_EVENT)
+        event_payload = new_tip_event['eventPayload']
+
+        assert_equal(event_payload['hash'], self.blocks[-1])
+        assert_equal(event_payload['height'], sc_last_we_block_height + 1)
+        assert_true('feePayments' not in event_payload)
+        expected_fee_payments = http_block_getFeePayments(sc_node, self.blocks[-1])['feePayments']
```
Comment (VolodymyrBabaiev) [inline qa/websocket_server_fee_payments.py:205]: Consider to rename expected_fee_payments to api_fee_payments.

File: qa/mc_sc_forging_fee_payments.py [hunk 2]
```diff
@@ -224,7 +225,24 @@ def run_test(self):
         assert_equal(sc_node2_balance_after_payments, sc_node2_balance_before_payments + node_2_fees,
                      "Wrong fee payment amount for SC node 2")
 
-
+        # Check fee payments from API perspective
+        # Non-last block of the epoch:
+        expected_fee_payments_node1 = http_block_getFeePayments(sc_node1, sc_middle_we_block_id)['feePayments']
+        assert_equal(0, len(expected_fee_payments_node1),
+                     "No fee payments expected to be found for the block in the middle of WE")
+        expected_fee_payments_node2 = http_block_getFeePayments(sc_node2, sc_middle_we_block_id)['feePayments']
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply qa/mc_sc_forging_fee_payments.py]: done

File: qa/websocket_server_fee_payments.py (hunk 1 above)
Comment (i-Alex) [reply qa/websocket_server_fee_payments.py]: done

File: sdk/src/main/scala/com/horizen/utils/FeePaymentsUtils.scala [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+object FeePaymentsUtils {
+  val DEFAULT_FEE_PAYMENTS_HASH: Array[Byte] = Utils.ZEROS_HASH
+
+  def calculateFeePaymentsHash(feePayments: Seq[ZenBox]): Array[Byte] = {
+    if(feePayments.isEmpty) {
+      // No fees for the whole epoch, so no fee payments for the Forgers.
+      DEFAULT_FEE_PAYMENTS_HASH
+    } else {
+      // TODO: create FeePaymentsTransaction and return its id() in bytes
+      val tx = new FeePaymentsTransaction(feePayments.asJava, FeePaymentsTransaction.FEE_PAYMENTS_TRANSACTION_VERSION)
```
Comment (alsala) [inline sdk/src/main/scala/com/horizen/utils/FeePaymentsUtils.scala:16]: Good to have a version

File: qa/mc_sc_forging_fee_payments.py [hunk 4]
```diff
@@ -224,7 +225,24 @@ def run_test(self):
         assert_equal(sc_node2_balance_after_payments, sc_node2_balance_before_payments + node_2_fees,
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply qa/mc_sc_forging_fee_payments.py:225]: Updated the test data to make this part valuable.

File: sdk/src/main/scala/com/horizen/SidechainState.scala [hunk 5]
```diff
@@ -174,10 +176,24 @@ class SidechainState private[horizen] (stateStorage: SidechainStateStorage,
...
+    // If SC block has reached the end of the withdrawal epoch -> fee payments expected to be produced.
+    // Verify that Forger assumed the same fees to be paid as the current node does.
+    // If SC block is in the middle of the withdrawal epoch -> no fee payments hash expected to be defined.
+    val isWithdrawalEpochFinished: Boolean = WithdrawalEpochUtils.isEpochLastIndex(modWithdrawalEpochInfo, params)
+    if(isWithdrawalEpochFinished) {
+      // Note: that current block fee info is still not in the state storage, so consider it during result calculation.
+      val feePayments = getFeePayments(modWithdrawalEpochInfo.epoch, Some(mod.feeInfo))
+      val feePaymentsHash: Array[Byte] = FeePaymentsUtils.calculateFeePaymentsHash(feePayments)
+
+      if(!mod.feePaymentsHash.sameElements(feePaymentsHash))
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/SidechainState.scala:191]: ok. Done in NodeViewHolder for every exception thrown by the state during block apply.

File: sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala [hunk 6]
```diff
@@ -268,12 +269,15 @@ class SidechainNodeViewHolder(sidechainSettings: SidechainSettings,
 
         updateInfo.state.applyModifier(modToApply) match {
           case Success(stateAfterApply) =>
-            val historyAfterApply = newHistory.reportModifierIsValid(modToApply)
+            var historyAfterApply = newHistory.reportModifierIsValid(modToApply)
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala]: doen

**Review by alsala — APPROVED** [review]

**Review by VolodymyrBabaiev — APPROVED** [review]

**Review by albertog78 — APPROVED** [review]

**Pull request merged**